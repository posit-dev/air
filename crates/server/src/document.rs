//! Types and utilities for working with documents

mod encoding;
mod key;
mod text_diff;
mod text_document;
mod text_edit;

pub(crate) use encoding::PositionEncoding;
pub(crate) use key::DocumentKey;
pub(crate) use text_document::DocumentVersion;
pub(crate) use text_document::TextDocument;
pub(crate) use text_edit::{Indel, TextEdit};
