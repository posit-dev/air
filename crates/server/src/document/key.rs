use url::Url;

/// A unique document ID, derived from a URL passed as part of an LSP request.
/// This document ID currently always points to an R file, but eventually can also
/// point to a full notebook, or a cell within a notebook.
#[derive(Clone, Debug)]
pub(crate) enum DocumentKey {
    Text(Url),
    // If we ever want to support notebooks, start here:
    // Notebook(Url),
    // NotebookCell(Url),
}

impl std::fmt::Display for DocumentKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Text(url) => url.fmt(f),
        }
    }
}
