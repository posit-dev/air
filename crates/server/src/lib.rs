//! ## The Air Language Server

pub use server::Server;

#[macro_use]
mod message;

mod crates;
mod document;
mod error;
mod logging;
mod proto;
mod server;
mod session;

#[cfg(test)]
mod test;

pub(crate) const SERVER_NAME: &str = "Air Language Server";
pub(crate) const SERVER_VERSION: &str = env!("CARGO_PKG_VERSION");
