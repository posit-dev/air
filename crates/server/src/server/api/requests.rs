mod format;
mod format_range;
mod view_file;

use super::traits::{BackgroundDocumentRequestHandler, RequestHandler};
pub(super) use format::Format;
pub(super) use format_range::FormatRange;
pub(super) use view_file::ViewFile;
