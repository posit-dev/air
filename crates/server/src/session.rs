//! Data model, state management, and configuration resolution.

use std::sync::Arc;

use lsp_types::Url;
use lsp_types::WorkspaceFolder;

use crate::edit::{DocumentKey, DocumentVersion, PositionEncoding, TextDocument};

pub(crate) use self::capabilities::ResolvedClientCapabilities;
pub use self::index::DocumentQuery;

mod capabilities;
mod index;
mod workspaces;

/// The global state for the LSP
pub(crate) struct Session {
    /// Used to retrieve information about open documents and settings.
    index: index::Index,
    /// The global position encoding, negotiated during LSP initialization.
    position_encoding: PositionEncoding,
    /// Tracks what LSP features the client supports and doesn't support.
    resolved_client_capabilities: Arc<ResolvedClientCapabilities>,
}

/// An immutable snapshot of `Session` that references
/// a specific document.
pub(crate) struct DocumentSnapshot {
    #[allow(dead_code)]
    resolved_client_capabilities: Arc<ResolvedClientCapabilities>,
    document_ref: index::DocumentQuery,
    position_encoding: PositionEncoding,
}

impl Session {
    pub(crate) fn new(
        resolved_client_capabilities: ResolvedClientCapabilities,
        position_encoding: PositionEncoding,
        workspace_folders: Vec<WorkspaceFolder>,
    ) -> crate::Result<Self> {
        Ok(Self {
            position_encoding,
            index: index::Index::new(workspace_folders)?,
            resolved_client_capabilities: Arc::new(resolved_client_capabilities),
        })
    }

    pub(crate) fn key_from_url(&self, url: Url) -> DocumentKey {
        self.index.key_from_url(url)
    }

    /// Creates a document snapshot with the URL referencing the document to snapshot.
    pub(crate) fn take_snapshot(&self, url: Url) -> Option<DocumentSnapshot> {
        let key = self.key_from_url(url);
        Some(DocumentSnapshot {
            resolved_client_capabilities: self.resolved_client_capabilities.clone(),
            document_ref: self.index.make_document_ref(key)?,
            position_encoding: self.position_encoding,
        })
    }

    /// Iterates over the LSP URLs for all open text documents. These URLs are valid file paths.
    #[allow(dead_code)]
    pub(crate) fn text_document_urls(&self) -> impl Iterator<Item = &lsp_types::Url> + '_ {
        self.index.text_document_urls()
    }

    /// Updates a text document at the associated `key`.
    ///
    /// The document key must point to a text document, or this will throw an error.
    pub(crate) fn update_text_document(
        &mut self,
        key: &DocumentKey,
        content_changes: Vec<lsp_types::TextDocumentContentChangeEvent>,
        new_version: DocumentVersion,
    ) -> crate::Result<()> {
        let encoding = self.encoding();

        self.index
            .update_text_document(key, content_changes, new_version, encoding)
    }

    /// Registers a text document at the provided `url`.
    /// If a document is already open here, it will be overwritten.
    pub(crate) fn open_text_document(&mut self, url: Url, document: TextDocument) {
        self.index.open_text_document(url, document);
    }

    /// De-registers a document, specified by its key.
    /// Calling this multiple times for the same document is a logic error.
    pub(crate) fn close_document(&mut self, key: &DocumentKey) -> crate::Result<()> {
        self.index.close_document(key)?;
        Ok(())
    }

    /// Reloads the settings index
    pub(crate) fn reload_settings(&mut self, changed_url: &Url) {
        self.index.reload_settings(changed_url);
    }

    /// Open a workspace folder at the given `url`.
    pub(crate) fn open_workspace_folder(&mut self, url: &Url) -> crate::Result<()> {
        self.index.open_workspace_folder(url)
    }

    /// Close a workspace folder at the given `url`.
    pub(crate) fn close_workspace_folder(&mut self, url: &Url) -> crate::Result<()> {
        self.index.close_workspace_folder(url)?;
        Ok(())
    }

    #[allow(dead_code)]
    pub(crate) fn num_documents(&self) -> usize {
        self.index.num_documents()
    }

    #[allow(dead_code)]
    pub(crate) fn num_workspaces(&self) -> usize {
        self.index.num_workspaces()
    }

    #[allow(dead_code)]
    pub(crate) fn resolved_client_capabilities(&self) -> &ResolvedClientCapabilities {
        &self.resolved_client_capabilities
    }

    pub(crate) fn encoding(&self) -> PositionEncoding {
        self.position_encoding
    }
}

impl DocumentSnapshot {
    #[allow(dead_code)]
    pub(crate) fn resolved_client_capabilities(&self) -> &ResolvedClientCapabilities {
        &self.resolved_client_capabilities
    }

    pub fn query(&self) -> &index::DocumentQuery {
        &self.document_ref
    }

    pub(crate) fn encoding(&self) -> PositionEncoding {
        self.position_encoding
    }
}
