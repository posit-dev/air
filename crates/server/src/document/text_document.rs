use biome_rowan::TextRange;
use lsp_types::TextDocumentContentChangeEvent;
use source_file::LineEnding;
use source_file::SourceFile;

use crate::document::PositionEncoding;
use crate::proto::TextRangeExt;

pub(crate) type DocumentVersion = i32;

/// The state of an individual document in the server. Stays up-to-date
/// with changes made by the user, including unsaved changes.
#[derive(Debug, Clone)]
pub struct TextDocument {
    /// The source file containing the contents and line index for the document.
    /// Line endings have been normalized to unix line endings here.
    source: SourceFile,
    /// The original line endings of the document. Used when sending changes back to the
    /// LSP client.
    ending: LineEnding,
    /// The latest version of the document, set by the LSP client. The server will panic in
    /// debug mode if we attempt to update the document with an 'older' version.
    version: DocumentVersion,
}

impl TextDocument {
    pub fn new(contents: String, version: DocumentVersion) -> Self {
        // Normalize to Unix line endings on the way in
        let (contents, ending) = source_file::normalize_newlines(contents);
        let source = SourceFile::new(contents);
        Self {
            source,
            ending,
            version,
        }
    }

    #[cfg(test)]
    pub fn doodle(contents: &str) -> Self {
        Self::new(contents.into(), 0)
    }

    #[cfg(test)]
    pub fn doodle_and_range(contents: &str) -> (Self, biome_text_size::TextRange) {
        let (contents, range) = crate::test::extract_marked_range(contents);
        let doc = Self::new(contents, 0);
        (doc, range)
    }

    pub fn contents(&self) -> &str {
        self.source.contents()
    }

    pub fn ending(&self) -> LineEnding {
        self.ending
    }

    pub fn source_file(&self) -> &SourceFile {
        &self.source
    }

    pub fn version(&self) -> DocumentVersion {
        self.version
    }

    pub fn apply_changes(
        &mut self,
        mut changes: Vec<lsp_types::TextDocumentContentChangeEvent>,
        new_version: DocumentVersion,
        encoding: PositionEncoding,
    ) {
        // Normalize line endings. Changing the line length of inserted or
        // replaced text can't invalidate the text change events, even those
        // applied subsequently, since those changes are specified with [line,
        // col] coordinates.
        for change in &mut changes.iter_mut() {
            let text = std::mem::take(&mut change.text);
            (change.text, _) = source_file::normalize_newlines(text);
        }

        if let [lsp_types::TextDocumentContentChangeEvent { range: None, .. }] = changes.as_slice()
        {
            tracing::trace!("Fast path - replacing entire document");
            // Unwrap: If-let ensures there is exactly 1 change event
            let change = changes.pop().unwrap();
            self.source = SourceFile::new(change.text);
            self.update_version(new_version);
            return;
        }

        for TextDocumentContentChangeEvent {
            range,
            text: change,
            ..
        } in changes
        {
            if let Some(range) = range {
                // Replace a range and rebuild the line index
                let range = TextRange::from_proto(range, &self.source, encoding);
                self.source.replace_range(
                    usize::from(range.start())..usize::from(range.end()),
                    &change,
                );
            } else {
                // Replace the whole file
                self.source = SourceFile::new(change);
            }
        }

        self.update_version(new_version);
    }

    fn update_version(&mut self, new_version: DocumentVersion) {
        let old_version = self.version;
        self.version = new_version;
        debug_assert!(self.version >= old_version);
    }
}

#[cfg(test)]
mod tests {
    use crate::document::{PositionEncoding, TextDocument};
    use lsp_types::{Position, TextDocumentContentChangeEvent};

    #[test]
    fn redo_edit() {
        let mut document = TextDocument::new(
            r#""""
测试comment
一些测试内容
"""
import click


@click.group()
def interface():
    pas
"#
            .to_string(),
            0,
        );

        // Add an `s`, remove it again (back to the original code), and then re-add the `s`
        document.apply_changes(
            vec![
                TextDocumentContentChangeEvent {
                    range: Some(lsp_types::Range::new(
                        Position::new(9, 7),
                        Position::new(9, 7),
                    )),
                    range_length: Some(0),
                    text: "s".to_string(),
                },
                TextDocumentContentChangeEvent {
                    range: Some(lsp_types::Range::new(
                        Position::new(9, 7),
                        Position::new(9, 8),
                    )),
                    range_length: Some(1),
                    text: String::new(),
                },
                TextDocumentContentChangeEvent {
                    range: Some(lsp_types::Range::new(
                        Position::new(9, 7),
                        Position::new(9, 7),
                    )),
                    range_length: Some(0),
                    text: "s".to_string(),
                },
            ],
            1,
            PositionEncoding::UTF16,
        );

        assert_eq!(
            document.contents(),
            r#""""
测试comment
一些测试内容
"""
import click


@click.group()
def interface():
    pass
"#
        );
    }

    #[test]
    fn test_document_position_encoding() {
        // Replace `b` after `𐐀` which is at position 5 in UTF-8
        let utf8_range = lsp_types::Range {
            start: lsp_types::Position {
                line: 0,
                character: 5,
            },
            end: lsp_types::Position {
                line: 0,
                character: 6,
            },
        };

        // `b` is at position 3 in UTF-16
        let utf16_range = lsp_types::Range {
            start: lsp_types::Position {
                line: 0,
                character: 3,
            },
            end: lsp_types::Position {
                line: 0,
                character: 4,
            },
        };

        let utf8_replace_params = vec![lsp_types::TextDocumentContentChangeEvent {
            range: Some(utf8_range),
            range_length: None,
            text: String::from("bar"),
        }];
        let utf16_replace_params = vec![lsp_types::TextDocumentContentChangeEvent {
            range: Some(utf16_range),
            range_length: None,
            text: String::from("bar"),
        }];

        let mut document = TextDocument::new("a𐐀b".into(), 1);
        document.apply_changes(
            utf8_replace_params,
            document.version + 1,
            PositionEncoding::UTF8,
        );
        assert_eq!(document.contents(), "a𐐀bar");

        let mut document = TextDocument::new("a𐐀b".into(), 1);
        document.apply_changes(
            utf16_replace_params,
            document.version + 1,
            PositionEncoding::UTF16,
        );
        assert_eq!(document.contents(), "a𐐀bar");
    }
}
