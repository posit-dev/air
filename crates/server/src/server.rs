//! Scheduling, I/O, and API endpoints.

use lsp_server as lsp;
use lsp_types as types;
use lsp_types::InitializeParams;
use std::num::NonZeroUsize;
// The new PanicInfoHook name requires MSRV >= 1.82
#[allow(deprecated)]
use std::panic::PanicInfo;
use types::DidChangeWatchedFilesRegistrationOptions;
use types::FileSystemWatcher;
use types::OneOf;
use types::TextDocumentSyncCapability;
use types::TextDocumentSyncKind;
use types::TextDocumentSyncOptions;
use types::WorkspaceFoldersServerCapabilities;

use self::connection::Connection;
use self::schedule::event_loop_thread;
use self::schedule::Scheduler;
use self::schedule::Task;
use crate::document::PositionEncoding;
use crate::message::try_show_message;
use crate::server::connection::ConnectionInitializer;
use crate::session::ResolvedClientCapabilities;
use crate::session::Session;

mod api;
mod client;
mod connection;
mod schedule;

pub(crate) use connection::ClientSender;

pub(crate) type Result<T> = std::result::Result<T, api::Error>;

pub struct Server {
    connection: Connection,
    client_capabilities: ResolvedClientCapabilities,
    worker_threads: NonZeroUsize,
    session: Session,
}

impl Server {
    pub fn new(
        worker_threads: NonZeroUsize,
        connection: lsp::Connection,
        connection_threads: Option<lsp::IoThreads>,
    ) -> anyhow::Result<Self> {
        let initializer = ConnectionInitializer::new(connection, connection_threads);

        let (id, initialize_params) = initializer.initialize_start()?;

        let client_capabilities = initialize_params.capabilities;
        let client_capabilities = ResolvedClientCapabilities::new(client_capabilities);
        let position_encoding = Self::find_best_position_encoding(&client_capabilities);
        let server_capabilities = Self::server_capabilities(position_encoding);

        let connection = initializer.initialize_finish(
            id,
            &server_capabilities,
            crate::SERVER_NAME,
            crate::SERVER_VERSION,
        )?;

        let InitializeParams {
            workspace_folders,
            client_info,
            ..
        } = initialize_params;

        let workspace_folders = workspace_folders.unwrap_or_default();

        // TODO: Get user specified options from `initialization_options`
        let log_level = None;
        let dependency_log_levels = None;

        crate::logging::init_logging(
            connection.make_sender(),
            log_level,
            dependency_log_levels,
            client_info,
        );

        crate::message::init_messenger(connection.make_sender());

        Ok(Self {
            connection,
            worker_threads,
            session: Session::new(
                client_capabilities.clone(),
                position_encoding,
                workspace_folders,
            )?,
            client_capabilities,
        })
    }

    pub fn run(self) -> anyhow::Result<()> {
        // The new PanicInfoHook name requires MSRV >= 1.82
        #[allow(deprecated)]
        type PanicHook = Box<dyn Fn(&PanicInfo<'_>) + 'static + Sync + Send>;
        struct RestorePanicHook {
            hook: Option<PanicHook>,
        }

        impl Drop for RestorePanicHook {
            fn drop(&mut self) {
                if let Some(hook) = self.hook.take() {
                    std::panic::set_hook(hook);
                }
            }
        }

        // unregister any previously registered panic hook
        // The hook will be restored when this function exits.
        let _ = RestorePanicHook {
            hook: Some(std::panic::take_hook()),
        };

        // When we panic, try to notify the client.
        std::panic::set_hook(Box::new(move |panic_info| {
            use std::io::Write;

            let backtrace = std::backtrace::Backtrace::force_capture();
            tracing::error!("{panic_info}\n{backtrace}");

            // we also need to print to stderr directly for when using `window/logMessage` because
            // the message won't be sent to the client.
            // But don't use `eprintln` because `eprintln` itself may panic if the pipe is broken.
            let mut stderr = std::io::stderr().lock();
            writeln!(stderr, "{panic_info}\n{backtrace}").ok();

            try_show_message(
                "The Air language server exited with a panic. See the logs for more details."
                    .to_string(),
                lsp_types::MessageType::ERROR,
            )
            .ok();
        }));

        event_loop_thread(move || {
            Self::event_loop(
                &self.connection,
                &self.client_capabilities,
                self.session,
                self.worker_threads,
            )?;
            self.connection.close()?;
            Ok(())
        })?
        .join()
    }

    #[allow(clippy::needless_pass_by_value)] // this is because we aren't using `next_request_id` yet.
    fn event_loop(
        connection: &Connection,
        resolved_client_capabilities: &ResolvedClientCapabilities,
        mut session: Session,
        worker_threads: NonZeroUsize,
    ) -> anyhow::Result<()> {
        let mut scheduler =
            schedule::Scheduler::new(&mut session, worker_threads, connection.make_sender());

        Self::try_register_capabilities(resolved_client_capabilities, &mut scheduler);
        for msg in connection.incoming() {
            if connection.handle_shutdown(&msg)? {
                break;
            }
            let task = match msg {
                lsp::Message::Request(req) => api::request(req),
                lsp::Message::Notification(notification) => api::notification(notification),
                lsp::Message::Response(response) => scheduler.response(response),
            };
            scheduler.dispatch(task);
        }

        Ok(())
    }

    fn try_register_capabilities(
        resolved_client_capabilities: &ResolvedClientCapabilities,
        scheduler: &mut Scheduler,
    ) {
        let _span = tracing::info_span!("try_register_capabilities").entered();

        // Register capabilities to the client
        let mut registrations: Vec<lsp_types::Registration> = vec![];

        if resolved_client_capabilities.dynamic_registration_for_did_change_watched_files {
            // Watch for changes in `air.toml` files so we can react dynamically
            let watch_air_toml_registration = lsp_types::Registration {
                id: String::from("air-toml-watcher"),
                method: "workspace/didChangeWatchedFiles".into(),
                register_options: Some(
                    serde_json::to_value(DidChangeWatchedFilesRegistrationOptions {
                        watchers: vec![FileSystemWatcher {
                            glob_pattern: lsp_types::GlobPattern::String("**/air.toml".into()),
                            kind: None,
                        }],
                    })
                    .unwrap(),
                ),
            };

            registrations.push(watch_air_toml_registration);
        } else {
            tracing::warn!("LSP client does not support watched files dynamic capability - automatic configuration reloading will not be available.");
        }

        if !registrations.is_empty() {
            let params = lsp_types::RegistrationParams { registrations };

            let response_handler = |()| {
                tracing::info!("Dynamic configuration successfully registered");
                Task::nothing()
            };

            if let Err(error) = scheduler
                .request::<lsp_types::request::RegisterCapability>(params, response_handler)
            {
                tracing::error!(
                    "An error occurred when trying to dynamically register capabilities: {error}"
                );
            }
        }
    }

    fn find_best_position_encoding(
        client_capabilities: &ResolvedClientCapabilities,
    ) -> PositionEncoding {
        // If the client supports UTF-8 we use that, even if it's not its
        // preferred encoding (at position 0). Otherwise we use the mandatory
        // UTF-16 encoding that all clients and servers must support, even if
        // the client would have preferred UTF-32. Note that VSCode and Positron
        // only support UTF-16.
        if client_capabilities
            .position_encodings
            .contains(&lsp_types::PositionEncodingKind::UTF8)
        {
            PositionEncoding::UTF8
        } else {
            PositionEncoding::UTF16
        }
    }

    fn server_capabilities(position_encoding: PositionEncoding) -> types::ServerCapabilities {
        types::ServerCapabilities {
            position_encoding: Some(position_encoding.into()),
            text_document_sync: Some(TextDocumentSyncCapability::Options(
                TextDocumentSyncOptions {
                    open_close: Some(true),
                    change: Some(TextDocumentSyncKind::INCREMENTAL),
                    will_save: Some(false),
                    will_save_wait_until: Some(false),
                    ..Default::default()
                },
            )),
            workspace: Some(types::WorkspaceServerCapabilities {
                workspace_folders: Some(WorkspaceFoldersServerCapabilities {
                    supported: Some(true),
                    change_notifications: Some(OneOf::Left(true)),
                }),
                file_operations: None,
            }),
            document_formatting_provider: Some(OneOf::Left(true)),
            document_range_formatting_provider: Some(OneOf::Left(true)),
            ..Default::default()
        }
    }
}
