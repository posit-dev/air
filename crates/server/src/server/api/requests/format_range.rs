use air_r_parser::RParserOptions;
use air_r_syntax::RExpressionList;
use air_r_syntax::RSyntaxKind;
use air_r_syntax::RSyntaxNode;
use biome_formatter::LineEnding;
use biome_rowan::AstNode;
use biome_rowan::Language;
use biome_rowan::SyntaxElement;
use biome_rowan::WalkEvent;
use biome_text_size::{TextRange, TextSize};
use lsp_types::{self as types, request as req, Range};
use workspace::settings::FormatSettings;

use crate::document::TextEdit;
use crate::document::{PositionEncoding, TextDocument};
use crate::proto::TextRangeExt;
use crate::server::api::LSPResult;
use crate::server::{client::Notifier, Result};
use crate::session::{DocumentQuery, DocumentSnapshot};

type FormatRangeResponse = Option<Vec<lsp_types::TextEdit>>;

pub(crate) struct FormatRange;

impl super::RequestHandler for FormatRange {
    type RequestType = req::RangeFormatting;
}

impl super::BackgroundDocumentRequestHandler for FormatRange {
    fn document_url(
        params: &types::DocumentRangeFormattingParams,
    ) -> std::borrow::Cow<lsp_types::Url> {
        std::borrow::Cow::Borrowed(&params.text_document.uri)
    }

    fn run_with_snapshot(
        snapshot: DocumentSnapshot,
        _notifier: Notifier,
        params: types::DocumentRangeFormattingParams,
    ) -> Result<FormatRangeResponse> {
        format_document_range(&snapshot, params.range)
    }
}

/// Formats the specified [`Range`] in the [`DocumentSnapshot`].
#[tracing::instrument(level = "info", skip_all)]
fn format_document_range(snapshot: &DocumentSnapshot, range: Range) -> Result<FormatRangeResponse> {
    let text_document = snapshot.query().as_single_document();
    let query = snapshot.query();
    format_text_document_range(text_document, range, query, snapshot.encoding())
}

/// Formats the specified [`Range`] in the [`TextDocument`].
fn format_text_document_range(
    text_document: &TextDocument,
    range: Range,
    query: &DocumentQuery,
    encoding: PositionEncoding,
) -> Result<FormatRangeResponse> {
    let document_settings = query.settings();
    let formatter_settings = &document_settings.format;

    let ending = text_document.ending();
    let source = text_document.source_file();
    let text = source.contents();
    let range = TextRange::from_proto(range, source, encoding);

    let Some((new_text, new_range)) = format_source_range(text, formatter_settings, range)
        .with_failure_code(lsp_server::ErrorCode::InternalError)?
    else {
        return Ok(None);
    };

    let text_edit = TextEdit::replace(new_range, new_text);

    let edits = text_edit
        .into_proto(source, encoding, ending)
        .with_failure_code(lsp_server::ErrorCode::InternalError)?;

    Ok(Some(edits))
}

fn format_source_range(
    source: &str,
    formatter_settings: &FormatSettings,
    range: TextRange,
) -> anyhow::Result<Option<(String, TextRange)>> {
    let parse = air_r_parser::parse(source, RParserOptions::default());

    if parse.has_errors() {
        return Err(anyhow::anyhow!("Can't format when there are parse errors."));
    }

    // Always use `Lf` line endings on the way out from the formatter since we
    // internally store all LSP text documents with `Lf` endings
    let format_options = formatter_settings
        .to_format_options(source)
        .with_line_ending(LineEnding::Lf);

    let logical_lines = find_deepest_enclosing_logical_lines(parse.syntax(), range);
    if logical_lines.is_empty() {
        tracing::warn!("Can't find logical line");
        return Ok(None);
    };

    // Find the overall formatting range by concatenating the ranges of the logical lines.
    // We use the "non-whitespace-range" as that corresponds to what Biome will format.
    let new_range = logical_lines
        .iter()
        .map(text_non_whitespace_range)
        .reduce(|acc, new| acc.cover(new))
        .expect("`logical_lines` is non-empty");

    // We need to wrap in an `RRoot` otherwise the comments get attached too
    // deep in the tree. See `CommentsBuilderVisitor` in biome_formatter and the
    // `is_root` logic. Note that `node` needs to be wrapped in at least two
    // other nodes in order to fix this problem, and here we have an `RRoot` and
    // `RExpressionList` that do the job.
    //
    // Since we only format logical lines, it is fine to wrap in an expression list.
    let Some(exprs): Option<Vec<air_r_syntax::AnyRExpression>> = logical_lines
        .into_iter()
        .map(air_r_syntax::AnyRExpression::cast)
        .collect()
    else {
        tracing::warn!("Can't cast to `AnyRExpression`");
        return Ok(None);
    };

    let list = air_r_factory::r_expression_list(exprs);
    let eof = air_r_syntax::RSyntaxToken::new_detached(RSyntaxKind::EOF, "", vec![], vec![]);
    let root = air_r_factory::r_root(list, eof).build();

    let printed = biome_formatter::format_sub_tree(
        root.syntax(),
        air_r_formatter::RFormatLanguage::new(format_options),
    )?;

    if printed.range().is_none() {
        // Happens in edge cases when biome returns a `Printed::new_empty()`
        return Ok(None);
    };

    let mut new_text = printed.into_code();

    // Remove last hard break line from our artifical expression list
    new_text.pop();

    Ok(Some((new_text, new_range)))
}

// From biome_formatter
fn text_non_whitespace_range<E, L>(elem: &E) -> TextRange
where
    E: Into<SyntaxElement<L>> + Clone,
    L: Language,
{
    let elem: SyntaxElement<L> = elem.clone().into();

    let start = elem
        .leading_trivia()
        .into_iter()
        .flat_map(|trivia| trivia.pieces())
        .find_map(|piece| {
            if piece.is_whitespace() || piece.is_newline() {
                None
            } else {
                Some(piece.text_range().start())
            }
        })
        .unwrap_or_else(|| elem.text_trimmed_range().start());

    let end = elem
        .trailing_trivia()
        .into_iter()
        .flat_map(|trivia| trivia.pieces().rev())
        .find_map(|piece| {
            if piece.is_whitespace() || piece.is_newline() {
                None
            } else {
                Some(piece.text_range().end())
            }
        })
        .unwrap_or_else(|| elem.text_trimmed_range().end());

    TextRange::new(start, end)
}

/// Finds consecutive logical lines. Currently that's only expressions at
/// top-level or in a braced list.
fn find_deepest_enclosing_logical_lines(node: RSyntaxNode, range: TextRange) -> Vec<RSyntaxNode> {
    let start_lists = find_expression_lists(&node, range.start(), false);
    let end_lists = find_expression_lists(&node, range.end(), true);

    // Both vectors of lists should have a common prefix, starting from the
    // program's expression list. As soon as the lists diverge we stop.
    let Some(list) = start_lists
        .into_iter()
        .zip(end_lists)
        .take_while(|pair| pair.0 == pair.1)
        .map(|pair| pair.0)
        .last()
    else {
        // Should not happen as the range is always included in the program's expression list
        tracing::warn!("Can't find common list parent");
        return vec![];
    };

    let Some(list) = RExpressionList::cast(list) else {
        tracing::warn!("Can't cast to expression list");
        return vec![];
    };

    let iter = list.into_iter();

    // We've chosen to be liberal about user selections and always widen the
    // range to include the selection bounds. If we wanted to be conservative
    // instead, we could use this `filter()` instead of the `skip_while()` and
    // `take_while()`:
    //
    // ```rust
    // .filter(|node| range.contains_range(node.text_trimmed_range()))
    // ```
    let logical_lines: Vec<RSyntaxNode> = iter
        .map(|expr| expr.into_syntax())
        .skip_while(|node| !node.text_range().contains(range.start()))
        .take_while(|node| node.text_trimmed_range().start() <= range.end())
        .collect();

    logical_lines
}

fn find_expression_lists(node: &RSyntaxNode, offset: TextSize, end: bool) -> Vec<RSyntaxNode> {
    let mut preorder = node.preorder();
    let mut nodes: Vec<RSyntaxNode> = vec![];

    while let Some(event) = preorder.next() {
        match event {
            WalkEvent::Enter(node) => {
                let Some(parent) = node.parent() else {
                    continue;
                };

                let is_contained = if end {
                    let trimmed_node_range = node.text_trimmed_range();
                    trimmed_node_range.contains_inclusive(offset)
                } else {
                    let node_range = node.text_range();
                    node_range.contains(offset)
                };

                if !is_contained {
                    preorder.skip_subtree();
                    continue;
                }

                if parent.kind() == RSyntaxKind::R_EXPRESSION_LIST {
                    nodes.push(parent.clone());
                    continue;
                }
            }

            WalkEvent::Leave(_) => {}
        }
    }

    nodes
}

#[cfg(test)]
mod tests {
    use crate::document::TextDocument;
    use crate::{test::init_test_client, test::TestClientExt};

    #[test]
    fn test_format_range_none() {
        let mut client = init_test_client();

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"<<>>",
        );

        let output = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"<<
>>",
        );

        let output = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"<<1
>>",
        );

        let output = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output);

        client.shutdown();
        client.exit();
    }

    #[test]
    fn test_format_range_logical_lines() {
        let mut client = init_test_client();

        // 2+2 is the logical line to format
        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"1+1
<<2+2>>
",
        );
        let output = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"1+1
#
<<2+2>>
",
        );

        let output = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output);

        // The element in the braced expression is a logical line
        // FIXME: Should this be the whole `{2+2}` instead?
        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"1+1
{<<2+2>>}
",
        );

        let output = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output);

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"1+1
<<{2+2}>>
",
        );
        let output = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output);

        // The deepest element in the braced expression is our target
        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"1+1
{
  2+2
  {
    <<3+3>>
  }
}
",
        );

        let output = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output);

        client.shutdown();
        client.exit();
    }

    #[test]
    fn test_format_range_mismatched_indent() {
        let mut client = init_test_client();

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"1
  <<2+2>>
",
        );

        // We don't change indentation when `2+2` is formatted
        let output = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output);

        // Debatable: Should we make an effort to remove unneeded indentation
        // when it's part of the range?
        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"1
<<  2+2>>
",
        );
        let output_wide = client.format_document_range(&doc, range);
        assert_eq!(output, output_wide);

        client.shutdown();
        client.exit();
    }

    #[test]
    fn test_format_range_multiple_lines() {
        let mut client = init_test_client();

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"1+1
<<#
2+2>>
",
        );

        let output1 = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output1);

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"<<1+1
#
2+2>>
",
        );
        let output2 = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output2);

        client.shutdown();
        client.exit();
    }

    #[test]
    fn test_format_range_unmatched_lists() {
        let mut client = init_test_client();

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"0+0
<<1+1
{
  2+2>>
}
3+3
",
        );

        let output1 = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output1);

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"0+0
<<1+1
{
>>  2+2
}
3+3
",
        );
        let output2 = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output2);

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"0+0
<<1+1
{
  2+2
}
>>3+3
",
        );
        let output3 = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output3);

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"0+0
1+1
{
<<  2+2
}
>>3+3
",
        );
        let output4 = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output4);

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"<<1+1>>
2+2
",
        );

        let output5 = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output5);

        #[rustfmt::skip]
        let (doc, range) = TextDocument::doodle_and_range(
"1+1
<<2+2>>
",
        );

        let output6 = client.format_document_range(&doc, range);
        insta::assert_snapshot!(output6);

        client.shutdown();
        client.exit();
    }
}
