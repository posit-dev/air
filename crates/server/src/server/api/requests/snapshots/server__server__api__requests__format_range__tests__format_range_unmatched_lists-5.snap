---
source: crates/server/src/server/api/requests/format_range.rs
expression: output5
---
1 + 1
2+2
